### ISSUE TITLE
Emit length check and Array.Empty when collection-expr may be empty

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#71194]
Related to #68785

For a scenario like `T[] arr = [..elem1, ..elem2]`, we should consider emitting a branch like the following:

```cs
var elem1Temp = elem1;
var elem2Temp = elem2;
var length = elem1Temp.Length + elem2Temp.Length;
T[] arr;
if (length == 0)
{
    arr = Array.Empty<T>();
}
else
{
    arr = new T[length];
    // ... populate the array
}
```

**Comment by CyrusNajmabadi** [comment]
Instead of emitting that at all sites, could you emit a helper you call?  Jit can decide to inline as appropriate.

**Comment by alrz** [comment]
>T[] arr;

Wouldn't populating a `Span<T>` and calling `ToArray` do the same thing?

**Comment by RikkiGibson** [comment]
Yes but I think using Span.ToArray isn't the right solution here. Stack alloc'd spans are only used by this feature when the length is a compile time constant. But length is not constant when spreads are used. If we used an array-backed span instead then we are incurring an extra allocation instead of avoiding one, and also incurring an extra value copy in either case.
